### ISSUE TITLE
covscan issue in xsk code

### ISSUE THREAD

**Issue by tohojo** [xdp-project/xdp-tools#117]
Hi @magnus-karlsson 

Covscan turned up this report in the XSK code - care to take a look?

```
*** CID 317140:  Null pointer dereferences  (REVERSE_INULL)
/lib/libbpf/src/xsk.c: 1029 in xsk_socket__create_shared()
1023     	struct xsk_socket *xsk;
1024     	struct xsk_ctx *ctx;
1025     	int err, ifindex;
1026     	bool unmap = umem->fill_save != fill;
1027     	bool rx_setup_done = false, tx_setup_done = false;
1028     
>>>     CID 317140:  Null pointer dereferences  (REVERSE_INULL)
>>>     Null-checking "umem" suggests that it may be null, but it has already been dereferenced on all paths leading to the check.
1029     	if (!umem || !xsk_ptr || !(rx || tx))
1030     		return -EFAULT;
1031     
1032     	xsk = calloc(1, sizeof(*xsk));
1033     	if (!xsk)
1034     		return -ENOMEM;
```

**Comment by magnus-karlsson** [comment]
This has already been fixed. We should probably update the libbpf in the libxdp repo.

/Magnus

commit afd0be7299533bb2e2b09104399d8a467ecbd2c5
Author: Ciara Loftus <ciara.loftus@intel.com>
Date:   Thu Apr 8 05:20:09 2021 +0000

    libbpf: Fix potential NULL pointer dereference
    
    Wait until after the UMEM is checked for null to dereference it.
    
    Fixes: 43f1bc1efff1 ("libbpf: Restore umem state after socket create failure")
    Signed-off-by: Ciara Loftus <ciara.loftus@intel.com>
    Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>
    Link: https://lore.kernel.org/bpf/20210408052009.7844-1-ciara.loftus@intel.com

diff --git a/tools/lib/bpf/xsk.c b/tools/lib/bpf/xsk.c
index d24b5cc720ec..007fe5d59438 100644
--- a/tools/lib/bpf/xsk.c
+++ b/tools/lib/bpf/xsk.c
@@ -852,18 +852,19 @@ int xsk_socket__create_shared(struct xsk_socket **xsk_ptr,
                              struct xsk_ring_cons *comp,
                              const struct xsk_socket_config *usr_config)
 {
+       bool unmap, rx_setup_done = false, tx_setup_done = false;
        void *rx_map = NULL, *tx_map = NULL;
        struct sockaddr_xdp sxdp = {};
        struct xdp_mmap_offsets off;
        struct xsk_socket *xsk;
        struct xsk_ctx *ctx;
        int err, ifindex;
-       bool unmap = umem->fill_save != fill;
-       bool rx_setup_done = false, tx_setup_done = false;
 
        if (!umem || !xsk_ptr || !(rx || tx))
                return -EFAULT;
 
+       unmap = umem->fill_save != fill;
+
        xsk = calloc(1, sizeof(*xsk));
        if (!xsk)
                return -ENOMEM;

**Comment by tohojo** [comment]
Ohh... totally missed that it was in the libbpf code; my bad! Will update that :)
